[WIP] Balanced Random Forest

massich opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

Fixes #8607 


#### What does this implement/fix? Explain your changes.
This PR takes over #5181

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

potash committed:
initial balanced commit

potash committed:
fix default value, comment encoding

potash committed:
remove debug

potash committed:
cache balance_data

potash committed:
subsetting data is more efficient

potash committed:
fix sample_weight when balanced

potash committed:
fix sample_weights

potash committed:
balanced random forest example

potash committed:
Raise error for balanced multi-output

potash committed:
remove brf example

potash committed:
refactor

potash committed:
multi-output brf

potash committed:
fix flake8

potash committed:
xrange -> range

potash committed:
flake8

potash commented:
@massich thanks so much for resurrecting this PR-- I'm still using the feature in multiple projects and so would love to see it incorporated. I just pushed some flake8 fixes to my original PR #5181.

What's your plan for tests? I have a set of real-world benchmarks in my [brf-example](https://github.com/potash/brf-example) repo. Let me know how I can help.

glemaitre commented:
> What's your plan for tests?

@potash We want to extend the benchmark on several imbalanced dataset.
In this regard, we will use the datasets which we recently added in the [`imbalanced-learning`](http://contrib.scikit-learn.org/imbalanced-learn/generated/imblearn.datasets.fetch_datasets.html#imblearn.datasets.fetch_datasets) as well as the additional metrics available.

> I'm still using the feature in multiple projects and so would love to see it incorporated. I just pushed some flake8 fixes to my original PR #5181.

Could you reset the history to avoid merging master into your branch. Rebasing should be the way.
In fact, an easy thing could be to checkout the current branch and cherry-pick your PEP8 commits and make a PR on @massich branch.

> Let me know how I can help.

I am sure that we will need support to make sense of the benchmarking results.

potash commented:
@glemaitre I cherry-picked the flake8 fixes and put them in a [PR](https://github.com/massich/scikit-learn/pull/2) to @masich.

massich committed:
Merge pull request #2 from potash/feature/balanced-random-forest

Feature/balanced random forest

Commit:
Set-up class_weight='balanced_bootstraping' to trigger BRF

Change `balance_data=Ture` for `class_weight='balanced_bootstraping'`. See #8607
discussion

massich commented:
This work is followed here #8732 'cos I messed up somehow and I don't find howto retake the PR
